OpenRA.Mods.Common/MapFormats/DefaultMap.cs

```diff
@@ -0,0 +1,536 @@
+		public readonly byte TileFormat = 2;
+
+		public static string ComputeUID(IReadOnlyPackage package)
+		{
+			return ComputeUID(package, GetMapFormat(package));
+		}
+
+		public static string ComputeUID(IReadOnlyPackage package, int format)
+		{
+			// UID is calculated by taking an SHA1 of the yaml and binary data
+			var requiredFiles = new[] { "map.yaml", "map.bin" };
+			var contents = package.Contents.ToList();
+			foreach (var required in requiredFiles)
+				if (!contents.Contains(required))
+					throw new FileNotFoundException($"Required file {required} not present in this map");
+
+			var streams = new List<Stream>();
+			try
+			{
+				foreach (var filename in contents)
+					if (filename.EndsWith(".yaml") || filename.EndsWith(".bin") || filename.EndsWith(".lua") || (format >= 12 && filename == "map.png"))
+						streams.Add(package.GetStream(filename));
+
+				// Take the SHA1
+				if (streams.Count == 0)
+					return CryptoUtil.SHA1Hash(Array.Empty<byte>());
+
+				var merged = streams[0];
+				for (var i = 1; i < streams.Count; i++)
+					merged = new MergedStream(merged, streams[i]);
+
+				return CryptoUtil.SHA1Hash(merged);
+			}
+			finally
+			{
+				foreach (var stream in streams)
+					stream.Dispose();
+			}
+		}
+
+		/// <summary>
+		/// Initializes a new map created by the editor or importer.
+		/// The map will not receive a valid UID until after it has been saved and reloaded.
+		/// </summary>
+		public DefaultMap(ModData modData, ITerrainInfo terrainInfo, int width, int height)
+			: base(modData)
+		{
+			var tileRef = terrainInfo.DefaultTerrainTile;
+			var size = new Size(width, height);
+
+			Title = "Name your map here";
+			Author = "Your name here";
+
+			MapSize = new int2(size);
+			Tileset = terrainInfo.Id;
+
+			NewSize(new Size(width, height), terrainInfo);
+
+			Tiles.Clear(tileRef);
+
+			PostInit();
+		}
+
+		public DefaultMap(ModData modData, IReadOnlyPackage package)
+			: base(modData)
+		{
+			Package = package;
+
+			if (!Package.Contains("map.yaml") || !Package.Contains("map.bin"))
+				throw new InvalidDataException($"Not a valid map\n File: {package.Name}");
+
+			var yaml = new MiniYaml(null, MiniYaml.FromStream(Package.GetStream("map.yaml"), package.Name));
+			foreach (var field in YamlFields)
+				field.Deserialize(this, yaml.Nodes);
+
+			if (MapFormat < SupportedMapFormat)
+				throw new InvalidDataException($"Map format {MapFormat} is not supported.\n File: {package.Name}");
+
+			PlayerDefinitions = MiniYaml.NodesOrEmpty(yaml, "Players");
+			ActorDefinitions = MiniYaml.NodesOrEmpty(yaml, "Actors");
+
+			Ruleset rules;
+			try
+			{
+				rules = Ruleset.Load(modData, this, Tileset, RuleDefinitions, WeaponDefinitions,
+					VoiceDefinitions, NotificationDefinitions, MusicDefinitions, ModelSequenceDefinitions);
+			}
+			catch (Exception e)
+			{
+				Log.Write("debug", "Failed to load rules for {0} with error {1}", Title, e);
+				rules = Ruleset.LoadDefaultsForTileSet(modData, Tileset);
```

Comment:
"Unnecessary assignment of a value to 'rules'"

Something went really wrong here. `rules` is set but not used